parachains/pallets/collective-content/src/lib.rs

```diff
@@ -0,0 +1,276 @@
+
+			let mut announcements = <Announcements<T, I>>::get();
+			announcements
+				.try_push((cid.clone(), maybe_expire_at.clone()))
+				.map_err(|_| Error::<T, I>::TooManyAnnouncements)?;
+			<Announcements<T, I>>::put(announcements);
+
+			if let Some(expire_at) = maybe_expire_at {
+				if NextAnnouncementExpireAt::<T, I>::get().map_or(true, |n| n > expire_at) {
+					NextAnnouncementExpireAt::<T, I>::put(expire_at);
+				}
+			}
+
+			Self::deposit_event(Event::<T, I>::AnnouncementAnnounced { cid, maybe_expire_at });
+			Ok(())
+		}
+
+		/// Remove an announcement.
+		///
+		/// Parameters:
+		/// - `origin`: Must be the [Config::CharterOrigin].
+		/// - `cid`: [CID](super::Cid) of the IPFS document to remove.
+		///
+		/// Weight: `O(1)`, less of the [Config::MaxAnnouncementsCount] is lower.
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::remove_announcement())]
+		pub fn remove_announcement(origin: OriginFor<T>, cid: Cid) -> DispatchResult {
+			T::AnnouncementOrigin::ensure_origin(origin)?;
+
+			let mut announcements = <Announcements<T, I>>::get();
+			let pos = announcements
+				.binary_search_by_key(&cid, |(c, _)| c.clone())
+				.ok()
+				.ok_or(Error::<T, I>::MissingAnnouncement)?;
+			announcements.remove(pos);
+			<Announcements<T, I>>::put(announcements);
+
+			Self::deposit_event(Event::<T, I>::AnnouncementRemoved { cid });
+			Ok(())
+		}
+	}
+
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// Clean up expired announcements.
+		pub fn cleanup_announcements(now: T::BlockNumber) {
+			if NextAnnouncementExpireAt::<T, I>::get().map_or(true, |next| next > now) {
+				// no expired announcements expected.
+				return
+			}
+			// find expired announcements.
+			let announcements = <Announcements<T, I>>::get();
+			let (live, expired): (Vec<(_, _)>, Vec<(_, _)>) =
+				announcements.clone().into_iter().partition(|(_, maybe_expire_at)| {
+					maybe_expire_at.map_or(true, |expire_at| expire_at > now)
+				});
+			if expired.len() == 0 {
+				// no expired announcements.
+				return
+			}
+			// convert into bounded vec.
+			let live: BoundedVec<(_, _), T::MaxAnnouncementsCount> = live.try_into().map_or_else(
+				|err| {
+					// should never happen.
+					debug_assert!(false, "failed to convert into bounded vec: {:?}", err);
+					announcements
+				},
+				|live| live,
+			);
+			// determine the next announcement expire at.
+			NextAnnouncementExpireAt::<T, I>::set(live.clone().into_iter().fold(
+				None,
+				|next, (_, maybe_expire_at)| match (next, maybe_expire_at) {
+					(Some(next), Some(expire_at)) if next > expire_at => Some(expire_at),
+					(None, Some(expire_at)) => Some(expire_at),
+					_ => next,
+				},
+			));
+			// save new list of announcements.
+			<Announcements<T, I>>::put(live);
+			// publish events for removed announcements.
+			expired.into_iter().for_each(|(cid, _)| {
+				Self::deposit_event(Event::<T, I>::AnnouncementRemoved { cid })
+			});
```

Comment:
Will a simple for suffice here:
```suggestion
			for (cid, _) in expired {
				Self::deposit_event(Event::<T, I>::AnnouncementRemoved { cid })
			}
```